File: src/main/java/com/pubnub/api/endpoints/access/AccessManagerService.java
package com.pubnub.api.endpoints.access;

import com.pubnub.api.models.server.access_manager.AccessManagerGrantPayload;
import com.pubnub.api.models.server.Envelope;
import com.pubnub.api.models.server.access_manager.AccessManagerAuditPayload;
import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.QueryMap;

import java.util.Map;

public interface AccessManagerService {

    @GET("/v1/auth/grant/sub-key/{subKey}")
    Call<Envelope<AccessManagerGrantPayload>> grant(@Path("subKey") String subKey, @QueryMap Map<String, String> options);

    @GET("/v1/auth/audit/sub-key/{subKey}")
    Call<Envelope<AccessManagerAuditPayload>> audit(@Path("subKey") String subKey, @QueryMap Map<String, String> options);

}

File: core/src/main/kotlin/com/pubnub/internal/services/AccessManagerService.kt
package com.pubnub.internal.services

import com.pubnub.internal.models.server.Envelope
import com.pubnub.internal.models.server.access_manager.AccessManagerGrantPayload
import com.pubnub.internal.models.server.access_manager.v3.GrantTokenResponse
import com.pubnub.internal.models.server.access_manager.v3.RevokeTokenResponse
import retrofit2.Call
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.QueryMap

internal interface AccessManagerService {

    @GET("/v2/auth/grant/sub-key/{subKey}")
    fun grant(
        @Path("subKey") subKey: String,
        @QueryMap options: Map<String, String>
    ): Call<Envelope<AccessManagerGrantPayload>>

    @POST("/v3/pam/{subKey}/grant")
    fun grantToken(
        @Path("subKey") subKey: String?,
        @Body body: Any?,
        @QueryMap options: Map<String, String>
    ): Call<GrantTokenResponse>

    @DELETE("/v3/pam/{subKey}/grant/{token}")
    fun revokeToken(
        @Path("subKey") subKey: String,
        @Path("token", encoded = true) token: String,
        @QueryMap queryParams: Map<String, String>
    ): Call<RevokeTokenResponse>
}
